### ISSUE TITLE
NOT NULL error

### ISSUE THREAD

**Issue by tonyukuk6** [sqlitebrowser/sqlitebrowser#3266]
Details for the issue
I think "not null" constraint is not working properly. I left a cell empty which is "NOT NULL" but it allowed to store data. I tried the same in SQLiteStudio software for test purpose. An error message (attached screenshot) appeared as expected.

#### What did you see instead?

Useful extra information
-------------------------
> DB4S v3.12.2 [built for x86_64-little_endian-llp64] on Windows 7 SP 1 (6.1) (winnt/6.1.7601) [x86_64]
> using SQLite Version 3.35.5
> and Qt 5.12.8
![Screenshot_1](https://user-images.githubusercontent.com/123555349/214556804-acb75361-49fc-4053-accd-1bd4b27efb22.png)

**Comment by chrisjlocke** [comment]
Do you have the actual 'create' statement for that table?
What data was you actually adding in?

> I left a cell empty which is "NOT NULL" but it allowed to store data. 

Can you define 'empty'?  While '' is an empty string, it is not NULL.  NULL is absolutely nothing.  So if you define a cell as 'not null', it can still contain an empty string.

You've provided a screenshot of some other program, but not with the issue with DB4S.  We may need that too, just to visualize what you're seeing - it's hard to know exactly what the problem is without it.

**Comment by scottfurry** [comment]
For clarity, see [SQL Anti-Patterns - Null Usage](https://github.com/boralp/sql-anti-patterns#null-usage).

Quoting...

> NULL is not the same as zero. A number ten greater than an unknown is still an unknown. NULL is not the same as a string of zero length. Combining any string with NULL in standard SQL returns NULL. NULL is not the same as false. Boolean expressions with AND, OR, and NOT also produce results that some people find confusing. When you declare a column as NOT NULL, it should be because it would make no sense for the row to exist without a value in that column. Use null to signify a missing value for any data type.

This is one of those *"missing nuggets of knowledge that will bite you in the backside"* that many using SQL fail to appreciate when taught.

**Comment by chrisjlocke** [comment]
Sidestepping slightly, but some languages don't like NULLs.  VB .net for instance (defaultly) doesn't like NULL integers or dates.  Create a date and slap nothing in it, and it'll default to 0001/01/01 00:00.  Thats not nothing (or even a really valid date) but its 'something'.  A null integer becomes 0.  So this can be confusing when reading NULL data from a database, and VB mangles it so its not null.  When writing that data back out ... bye bye NULLs.  But I digress. *

*VB now has nullable types, but you have to declare them as so, so if you forget, your butt gets bitten.

**Comment by RafaelMosca** [comment]
I think the confusing part (at least to me) is that, hitting delete in a table editor, sets the cell as empty string.

But there is a right click menu to set the cell to null :)

**Comment by scottfurry** [comment]
> I think the confusing part (at least to me) is that, hitting delete in a table editor, sets the cell as empty string.
> 
> But there is a right click menu to set the cell to null :)

... and since GUI software doesn't really like null values ( i.e. `nullptr` for C++ or `None` for python) it populates the control with a default value for a default type. The empty string is not null.

**Comment by scottfurry** [comment]
Think of it this way - when adding new data, does a value HAVE TO BE PRESENT?
If the answer is yes, field is marked `NOT NULL`.

And in SQL or other code, you need to ensure a value is defined in the `INSERT` statement.
Hypothetical:
```
INSERT INTO 'sometable' VALUES (PK, 'fieldA', 'fieldB');

INSERT INTO 'sometable' VALUES (PK, 'fieldA');
```
If the table definition for `sometable` has `fieldB` marked as NOT NULL, the second bogus insert will fail.

GUI software cannot necessarily handle the absence of data notion very well. And leaving null pointers lying around code is a disaster waiting to happen. A default value type is assumed and set to empty (i.e. NUMERIC -> 0, DECIMAL -> 0.0, VARCHAR -> empty string). And the default value is, by definition in SQL, `NOT NULL`.

Bottom line... GUI software and SQL don't mix nicely on occasion.
